src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Tuples.cs

```diff
@@ -0,0 +1,353 @@
+                validator.ValidateAttributesOnProperties();
+                validator.ValidateAttributesOnIndexer();
+            }
+
+            private void ValidateAttributesOnNamedTypes()
+            {
+                // public class Base0 { }
+                ValidateTupleNameAttribute(_base0Class, expectedTupleNamesAttribute: false);
+
+                // public class Base1<T> { }
+                ValidateTupleNameAttribute(_base1Class, expectedTupleNamesAttribute: false);
+
+                // public class Base2<T, U> { }
+                ValidateTupleNameAttribute(_base2Class, expectedTupleNamesAttribute: false);
+
+                // public class Outer<T> : Base1<(int key, int val)>
+                Assert.True(_outerClass.BaseType.ContainsTuple());
+                var expectedElementNames = new[] { "key", "val" };
+                ValidateTupleNameAttribute(_outerClass,
+                    expectedTupleNamesAttribute: true,
+                    expectedElementNames: expectedElementNames);
+
+                // public class Derived<T> : Outer<(int e1, (int e2, int e3) e4)>.Inner<
+                //     Outer<(int e5, int e6)>.Inner<(int e7, int e8)[], (int e9, int e10)>.InnerInner<int>[],
+                //     ((int e11, int e12) e13, int e14)>
+                //     .InnerInner<((int e15, int e16) e17, (int e18, Base1<(int e19, int e20)> e21) e22)>
+                expectedElementNames = new[] {
+                    "e1", "e4", "e2", "e3", "e5", "e6", "e7", "e8", "e9",
+                    "e10", "e13", "e14", "e11", "e12", "e17", "e22", "e15",
+                    "e16", "e18", "e21", "e19", "e20"
+                };
+                ValidateTupleNameAttribute(_derivedClass,
+                    expectedTupleNamesAttribute: true,
+                    expectedElementNames: expectedElementNames);
+            }
+
+            private void ValidateAttributesOnFields()
+            {
+                // public static (int e1, int e2) Field1;
+                var field1 = _derivedClass.GetMember<FieldSymbol>("Field1");
+                var expectedElementNames = new[] { "e1", "e2" };
+                ValidateTupleNameAttribute(field1, expectedTupleNamesAttribute: true, expectedElementNames: expectedElementNames);
+
+                // public static (dynamic e1, int e2) Field2;
+                var field2 = _derivedClass.GetMember<FieldSymbol>("Field2");
+                expectedElementNames = new[] { "e1", "e2" };
+                ValidateTupleNameAttribute(field2, expectedTupleNamesAttribute: true, expectedElementNames: expectedElementNames);
+
+                // public static Base1<(dynamic e1, (int e2, int e3))> Field3;
+                var field3 = _derivedClass.GetMember<FieldSymbol>("Field3");
+                expectedElementNames = new[] { "e1", "e4", "e2", "e3" };
+                ValidateTupleNameAttribute(field3, expectedTupleNamesAttribute: true, expectedElementNames: expectedElementNames);
```

Comment:
`Field2` and `Field3` are not `dynamic`. Are we testing the `dynamic` case to ensure we're emitting `[Dynamic(...)]` as well as `[TupleNames(...)]`?